lib/fluent/plugin/kubernetes_metadata_watch_pods.rb

```diff
@@ -25,8 +25,20 @@ module WatchPods
 
     def start_pod_watch
       begin
-        resource_version = @client.get_pods.resourceVersion
-        watcher          = @client.watch_pods(resource_version)
+        options = {
+          resource_version: '0'  # Fetch from API server.
```

Comment:
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#list-pod-v1-core
```
resourceVersion | ... - if it's 0,  then we simply return what we currently have in cache, no guarantee;
```
so it is fast, but "unsafe"
I don't think we want to hardcode this value to `'0'` - it should at least be settable by the user, in case there are users who must have up-to-date data, or who want to use whatever is the cluster policy e.g.
```
When specified for list: - if unset, then the result is returned from remote storage based on quorum-read flag;
```
I'm assuming `get_pods` is doing a `list` command.  So I think the user should have the option of either setting it to `'0'`, or omitting `resource_version` entirely from the query.

Reply:
@wojtek-t can you please comment on the implications of using the cache, especially on data freshness and how a user would choose whether to set `resourceVersion` to 0 or leave it unset?

@richm how should I make this configurable? A plug-in option?

Reply:
We don't provide explicit guarantee - that's correct.
But in practice, there is guarantee - as an example, we have 5s pod startup SLO: https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md
Why is that related? Basically starting a pod requires a couple rounds of watch, which is served exactly from the same cache.
So practice, in order to meet this SLO, it has to be in the order of tens-hundreds milliseconds, and that is what is happening.

Also, *all* Kubernetes component are working in that mode (e.g. it's build in the the reflector and informer frameworks). Kubelet, scheduler, controllers being part of controller-manager, ... all operate in that mode:
- list from cache: with RV=0
- start watching from that point
- reconcile state based on that

So you're not introducing anything new - I would even say "you're very late in this game".
I don't consider it unsafe in any means.

Reply:
> I would even say "you're very late in this game".

So all "officially supported" kubernetes api watch clients use `resourceVersion: 0`?  @bparees do you know anything about this?  Is this what the operator sdk uses to set up watches?  Who would we talk to in the operator team about this?  I don't think it is prudent to be in the situation where we are the only openshift component using watches in this way.  I did a quick scan through the cluster-logging-operator vendored code and it isn't clear to me that it uses watches with `resourceversion: 0`

Reply:
To be more specific, the LIST request uses RV=0. The WATCH uses the RV in the response from LIST. This client already has the intended behavior for WATCH. The new behavior is to set RV=0 in the LIST request.

Reply:
@richm it's safe elsewhere because we are always "list/watching" (list what's there, and then watch for changes).  If you're just listing (and thus not going to then watch for/consume changes that happened after the version you got back in the list) then i would agree that it's unsafe.  i'm not sure which pattern is being used here?

Reply:
another note is that there's basically no difference between "we asked for the latest and got it, but then 2ms later the resource was updated" and "we asked for the cached version and got something 2ms out of date".   W/o more context on exactly what we're doing w/ this value it's hard for me to comment intelligently, but i'd be wary of any thought process which is "we have to get the latest value when we invoke the get because we can't tolerate getting back a slightly stale value".

Reply:
Correct - I was talking about exactly this case which is "list + watch". I agree that if you need strong consistency or just doing LIST without watching after it, it may not always be safe.